scripts/o3de/tests/test_repo.py

```diff
@@ -0,0 +1,205 @@
+
+@pytest.mark.usefixtures('init_register_repo_data')
+class TestRepos:
+    created_files = []
+    valid_urls = [
+        'http://o3derepo.org/repo.json',
+        'http://o3derepo.org/TestGem/gem.json',
+        'http://o3derepo.org/TestProject/project.json',
+        'http://o3derepo.org/TestTemplate/template.json',
+        'http://o3derecursiverepo.org/repo.json'
+    ]
+
+    @pytest.mark.parametrize("repo_path, expected_manifest_file, expected_result", [
+                                 pytest.param('http://o3de.org', pathlib.PurePath('o3de_manifest.json'), True),
+                                 pytest.param('http://o3de.org/incorrect', pathlib.PurePath('o3de_manifest.json'), False),
+                             ])
+    def test_get_repository_list(self, repo_path, expected_manifest_file,expected_result):
+        self.o3de_manifest_data = json.loads(TEST_O3DE_MANIFEST_JSON_PAYLOAD)
+
+        def load_o3de_manifest(manifest_path: pathlib.Path = None) -> dict:
+            return self.o3de_manifest_data
+
+        with patch('o3de.manifest.load_o3de_manifest', side_effect=load_o3de_manifest) as _1:
+            assert (repo_path in manifest.get_manifest_repos()) == expected_result
+
+    @pytest.mark.parametrize("repo_uri, expected_result, expected_in_repo, download_repo_data, created_file", [
+                                 pytest.param('http://o3derepo.org', 0, True, TEST_O3DE_REPO_JSON_PAYLOAD, TEST_O3DE_REPO_FILENAME),
+                                 pytest.param('http://o3derepo.org', 1, False, TEST_O3DE_REPO_BROKEN_JSON_PAYLOAD, ''),
+                                 pytest.param('http://o3derepo.org', 0, True, TEST_O3DE_REPO_WITH_OBJECTS_JSON_PAYLOAD, TEST_O3DE_REPO_GEM_FILE_NAME),
+                                 pytest.param('http://o3de.org/incorrect', 1, False, 0, ''),
+                                 pytest.param('http://o3derecursiverepo.org', 0,True, TEST_O3DE_RECURSIVE_REPO_JSON_PAYLOAD, TEST_O3DE_REPO_FILENAME),
+                             ])
+    def test_add_repository(self, repo_uri, expected_result, expected_in_repo, download_repo_data, created_file):
+
+        self.o3de_manifest_data = json.loads(TEST_O3DE_MANIFEST_JSON_PAYLOAD)
+        self.created_files.clear()
+
+        def load_o3de_manifest(manifest_path: pathlib.Path = None) -> dict:
+            return copy.deepcopy(self.o3de_manifest_data)
+
+        def save_o3de_manifest(manifest_data: dict, manifest_path: pathlib.Path = None) -> bool:
+            self.o3de_manifest_data = manifest_data
+            return True
+
+        def mocked_requests_get(url):
+            if url in self.valid_urls:
+                cm = MagicMock()
+                cm.getcode.return_value = 200
+                cm.read.return_value = 0
+                cm.__enter__.return_value = cm
+            else:
+                raise urllib.error.HTTPError(url, 404, "Not found", {}, 0)
+
+            return cm
```

Comment:
the variable name `cm` does not follow our best practices for variable names - in this case I'm guessing it means 'custom mock?'

Reply:
Yes, will update